lib/plugin-installer.js

```diff
@@ -0,0 +1,272 @@
+        env: env || process.env,
+        stdio: ['ignore', 'pipe', 'inherit']
+      };
+
+      const npmInstall = spawn('npm', installArgs, installOpts);
+
+      npmInstall.on('error', _ => {
+        reject(new Error('Cannot install', packageSpecifier));
+      });
+
+      const bufs = [];
+      let len = 0;
+      npmInstall.stdout.on('readable', () => {
+        const buf = npmInstall.stdout.read();
+
+        if (buf) {
+          bufs.push(buf);
+          len += buf.length;
+        }
+      });
+
+      npmInstall.stdout.on('end', () => {
+        const lines = Buffer.concat(bufs, len)
+          .toString()
+          .trim()
+          .split('\n');
+
+        const line = lines[lines.length - 1];
+
+        if (line.indexOf('\t') > -1) {
+          // npm >= 5
+          const output = lines[lines.length - 1].split('\t');
+
+          if (output.length < 4) {
+            reject(new Error('Cannot parse npm output while installing plugin.'));
+            return;
+          }
+
+          this.packageName = output[1];
+          pluginPath = path.join(cwd, output[3]);
+        } else {
+          // npm < 5
+          this.packageName = path.basename(line);
+          pluginPath = line;
+        }
+      });
+
+      npmInstall.on('exit', () => {
+        if (pluginPath) {
+          this.pluginManifest = require(pluginPath);
+
+          resolve({
+            packageName: this.packageName,
+            pluginManifest: this.pluginManifest
+          });
+        }
+      });
+    });
+  }
+
+  get existingPluginOptions () {
+    const config = this.config || require('./config');
+    const systemConfig = config.systemConfig;
+
+    const name = this.pluginKey;
+
+    const existingPluginOptions =
+      systemConfig.plugins && systemConfig.plugins[name]
+        ? systemConfig.plugins[name] : {};
+
+    return existingPluginOptions;
+  }
+
+  get pluginKey () {
+    let name = this.pluginManifest.name || this.packageName;
+    let abbreviatedPackageName = null;
+
+    if (!this.pluginManifest.name &&
+      this.packageName.startsWith(PluginInstaller.PACKAGE_PREFIX)) {
+      abbreviatedPackageName =
+        this.packageName.substr(PluginInstaller.PACKAGE_PREFIX.length);
+      name = abbreviatedPackageName;
+    }
+
+    return name;
+  }
+
+  updateConfigurationFiles ({
+    pluginOptions,
+    enablePlugin,
+    addPoliciesToWhitelist }) {
+    // WARNING (kevinswiber): Updating YAML while maintaining presentation
+    // style is not easy.  We're using the YAWN library here, which has
+    // a decent approach given the current state of available YAML parsers,
+    // but it's far from perfect.  Take a look at existing YAWN issues
+    // before making any optimizations.  If any section of this code looks
+    // ugly or inefficient, it may be that way for a reason (or maybe not).
+    //
+    // ¯\_(ツ)_/¯
+    //
+    // https://github.com/mohsen1/yawn-yaml/issues
+
+    return new Promise((resolve, reject) => {
```

Comment:
I do not see any asynchronous flow here. If I am correct (the method is huge, I might me missing it) why would you return a `Promise`?